File: 221118-dojo/app/src/main/java/Solution.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        int hit = 1;
        int miss = 5;

       if (cacheSize == 0) {
           return cities.length * miss;
       }

       int answer = 0;

       LinkedList<String> cache = new LinkedList<>();

       for (int i = 0; i < cities.length; i += 1) {
           String convertedWord = cities[i].toLowerCase();

           if (cache.remove(convertedWord)) {
               answer += hit;
               cache.add(convertedWord);
               continue;
           }

           if (!cache.remove(convertedWord)) {
               answer += miss;
               if (cache.size() >= cacheSize) {
                   cache.remove(0);
               }
               cache.add(convertedWord);
           }
       }
       return answer;
    }
}

File: 221118-dojo/javascript/solution.js
export default function solution(cacheSize, cities) {
  let answer = 0;
  const cache = new Map();

  cities.forEach((city) => {
    city = city.toLowerCase();

    if (cache.has(city)) {
      cache.delete(city);
      cache.set(city);
      answer += 1;

      return;
    }

    if (cache.size === cacheSize) {
      cache.delete(cache.keys().next().value);
    }

    if (cacheSize > 0) {
      cache.set(city);
    }

    answer += 5;
  });

  return answer;
}
